src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -878,16 +881,24 @@ VideoDecoder::DecodedOutput VideoDecoder::convertAVFrameToDecodedOutput(
 
 void VideoDecoder::convertAVFrameToDecodedOutputOnCPU(
     VideoDecoder::RawDecodedOutput& rawOutput,
-    DecodedOutput& output) {
+    DecodedOutput& output,
+    std::optional<torch::Tensor> preAllocatedOutputTensor) {
   int streamIndex = rawOutput.streamIndex;
   AVFrame* frame = rawOutput.frame.get();
   auto& streamInfo = streams_[streamIndex];
   if (output.streamType == AVMEDIA_TYPE_VIDEO) {
     if (streamInfo.colorConversionLibrary == ColorConversionLibrary::SWSCALE) {
-      int width = streamInfo.options.width.value_or(frame->width);
-      int height = streamInfo.options.height.value_or(frame->height);
-      torch::Tensor tensor = torch::empty(
-          {height, width, 3}, torch::TensorOptions().dtype({torch::kUInt8}));
+      torch::Tensor tensor;
+      if (preAllocatedOutputTensor.has_value()) {
+        // TODO: check shape of preAllocatedOutputTensor?
```

Comment:
I think we should TORCH_CHECK for height, width, shape, etc. here.

Reply:
I have this a try thinking it would be a simple assert like

```py
assert `shape[-3] == H, W, 3`
```
 
But it turns out it's not as simple. Some tensors come as HWC while some other come pas HWC. This is because the pre-allocated batched tensors are allocated as such:

https://github.com/pytorch/torchcodec/blob/c6a0a5a079da408df56fd7c06e3b801cbada4db1/src/torchcodec/decoders/_core/VideoDecoder.cpp#L171-L186

It then me realize that everything works, but it's pretty magical. We end up doing the `.pemute()` calls in different places, but I think it would be a lot cleaner if we allocated batched output only as `NHWC`, and then permute this entire `NHWC` tensor in one go. What we do right now is that we permute all the N `HWC` tensors, and that's probably not as efficient (or as clean).

I want to fix this as an immediate follow-up if that's OK. I gave it a try here, but it's not trivial and it might be preferable not to overcomplexify this PR.

Reply:
I'm in favor of @NicolasHug suggestion. The logic he points out is legacy from way back when, and it wasn't necessarily throught through in terms of long term maintenance and code health. Always doing it one way, and then permuting as needed on the way out, sounds easier and cleaner.

Reply:
Sounds good to me